Tasks/VsTest/vstest.ts

```diff
@@ -0,0 +1,515 @@
+        getTestResultsDirectory(runSettingsFile, path.join(workingDirectory, 'TestResults')).then(function(resultsDirectory) {
+            invokeVSTest(resultsDirectory).then(function(code) {
+                try {
+                    publishTestResults(resultsDirectory);
+                    tl.setResult(code, tl.loc('VstestReturnCode', code));
+                }
+                catch (error) {
+                    tl._writeLine("##vso[task.logissue type=error;code=" + error + ";TaskName=VSTest]");
+                    throw error;
+                }
+            })
+                .fail(function(err) {
+                    tl._writeLine("##vso[task.logissue type=error;code=" + err + ";TaskName=VSTest]");
+                    throw err;
+                });
+        });
+    }
+    else {
+        tl._writeLine("##vso[task.logissue type=warning;code=002004;]");
+        tl.warning("No test assemblies found matching the pattern: " + testAssembly);
+    }
+}
+catch (error) {
+    tl._writeLine("##vso[task.logissue type=error;code=" + error + ";TaskName=VSTest]");
+    throw error;
+}
+
+function getTestAssemblies(): Set<string> {
+    var testAssemblyFiles = [];
+    if (testAssembly.indexOf('*') >= 0 || testAssembly.indexOf('?') >= 0) {
+        tl.debug('Pattern found in solution parameter.');
+        var excludeTestAssemblies = [];
+        var allFiles = tl.find(sourcesDirectory);
+        var testAssemblyFilters = testAssembly.split(';');
+        testAssemblyFilters.forEach(function(testAssemblyFilter) {
+            if (testAssemblyFilter.startsWith("-:")) {
+                if (testAssemblyFilter.indexOf('*') >= 0 || testAssemblyFilter.indexOf('?') >= 0) {
+                    excludeTestAssemblies = excludeTestAssemblies.concat(getFilteredFiles(testAssemblyFilter.substr(2), allFiles));
+                }
+                else {
+                    excludeTestAssemblies.push(testAssemblyFilter.substr(2));
+                }
+            }
+            else if (testAssemblyFilter.indexOf('*') >= 0 || testAssemblyFilter.indexOf('?') >= 0) {
+                testAssemblyFiles = testAssemblyFiles.concat(getFilteredFiles(testAssemblyFilter, allFiles));
+            }
+            else {
+                testAssemblyFiles.push(testAssemblyFilter);
+            }
+        });
+        testAssemblyFiles = testAssemblyFiles.filter(x=> excludeTestAssemblies.indexOf(x) < 0);
+    }
+    else {
+        tl.debug('No Pattern found in solution parameter.');
+        var assemblies = testAssembly.split(';');
+        assemblies.forEach(function(assembly) {
+            testAssemblyFiles.push(assembly);
+        });
+    }
+    return new Set(testAssemblyFiles);
+}
+
+function invokeVSTest(testResultsDirectory: string): Q.Promise<number> {
+    var defer = Q.defer<number>();
+    if (vsTestVersion == "latest") {
+        vsTestVersion = null;
+    }
+    overrideTestRunParametersIfRequired(runSettingsFile).then(function(overriddenSettingsFile) {
```

Comment:
call it sync then if it's required. We don't have to write .then for functions which are supposed to be sync